cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java

```diff
@@ -53,16 +54,15 @@
    *
    * @param responder the responder used for sending response back to client
    * @param delegate the {@link HttpContentConsumer} to delegate calls to
-   * @param datasetContext context for accessing dataset inside transaction
-   * @param txContext the {@link TransactionContext} for creating transaction
+   * @param serviceContext the {@link TransactionalHttpServiceContext} for this handler.
    */
   BodyConsumerAdapter(DelayedHttpServiceResponder responder, HttpContentConsumer delegate,
-                      DatasetContext datasetContext, TransactionContext txContext) {
+                      TransactionalHttpServiceContext serviceContext) {
     this.responder = responder;
     this.delegate = delegate;
-    this.txContext = txContext;
-    this.datasetContext = datasetContext;
-    this.transactional = createTransactional(txContext, datasetContext);
+    this.txContext = serviceContext.newTransactionContext();
```

Comment:
I think the creation should moved into `txExecute` so that a new context is being creating for the transactional execution and get dismissed after use.